File: app/src/main/java/com/zhiyong/tingxie/QuizRepository.java
package com.zhiyong.tingxie;

import android.app.Application;
import androidx.lifecycle.LiveData;
import android.os.AsyncTask;
import android.util.Log;

import com.zhiyong.tingxie.db.Question;
import com.zhiyong.tingxie.db.Quiz;
import com.zhiyong.tingxie.db.QuizPinyin;
import com.zhiyong.tingxie.db.Word;
import com.zhiyong.tingxie.ui.main.QuizItem;
import com.zhiyong.tingxie.ui.word.WordItem;

import java.util.List;

/* A Repository is a class that abstracts access to multiple data sources.
A Repository manages query threads and allows you to use multiple backends.
In the most common example, the Repository implements the logic for deciding whether to fetch data
from a network or use results cached in the local database. */
public class QuizRepository {

    private static final String TAG = "QuizRepository";

    private QuizDao mQuizDao;
    private LiveData<List<QuizItem>> mAllQuizItems;
    private LiveData<QuizItem> mQuizItem;
    private LiveData<List<WordItem>> mWordItems;
    private LiveData<List<QuizPinyin>> mAllQuizPinyins;
    private LiveData<List<Question>> mAllQuestions;
    private LiveData<List<WordItem>> mRemainingQuestions;

    public QuizRepository(Application application, long quizId) {
        PinyinRoomDatabase db = PinyinRoomDatabase.getDatabase(application);
        mQuizDao = db.pinyinDao();
        Log.d(TAG, "QuizRepository: ");

        mAllQuizItems = mQuizDao.getAllQuizItems();
        mQuizItem = mQuizDao.getQuizItem(quizId);
        mWordItems = mQuizDao.getWordItemsOfQuiz(quizId);
        mAllQuizPinyins = mQuizDao.getAllQuizPinyins();
        mAllQuestions = mQuizDao.getAllQuestions();
        mRemainingQuestions = mQuizDao.getRemainingQuestions(quizId);
    }

    public LiveData<List<QuizItem>> getAllQuizItems() {
        return mAllQuizItems;
    }

    public LiveData<QuizItem> getQuizItem() {
        return mQuizItem;
    }

    public LiveData<List<WordItem>> getWordItemsOfQuiz() {
        return mWordItems;
    }

    public LiveData<List<QuizPinyin>> getAllQuizPinyins() {
        return mAllQuizPinyins;
    }

    public LiveData<List<Question>> getAllQuestions() {
        return mAllQuestions;
    }

    public LiveData<List<WordItem>> getRemainingQuestionsOfQuiz() {
        return mRemainingQuestions;
    }

    public void insertQuiz(Quiz quiz) {
        new insertQuizAsyncTask(mQuizDao).execute(quiz);
    }

    private static class insertQuizAsyncTask extends AsyncTask<Quiz, Void, Void> {
        private QuizDao mAsyncTaskDao;

        insertQuizAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final Quiz... params) {
            mAsyncTaskDao.insert(params[0]);
            return null;
        }
    }

    public void updateQuiz(Quiz quiz) {
        new updateQuizAsyncTask(mQuizDao).execute(quiz);
    }

    private static class updateQuizAsyncTask extends AsyncTask<Quiz, Void, Void> {
        private QuizDao mAsyncTaskDao;

        updateQuizAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(Quiz... quizzes) {
            mAsyncTaskDao.update(quizzes[0]);
            return null;
        }
    }

    public void insertQuestion(Question question) {
        new insertQuestionAsyncTask(mQuizDao).execute(question);
    }

    private static class insertQuestionAsyncTask extends AsyncTask<Question, Void, Void> {
        private QuizDao mAsyncTaskDao;

        insertQuestionAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(Question... params) {
            mAsyncTaskDao.insert(params[0]);
            return null;
        }
    }

    public void updateQuestions(Long quizId) {
        new updateQuestionsAsyncTask(mQuizDao).execute(quizId);
    }

    private static class updateQuestionsAsyncTask extends AsyncTask<Long, Void, Void> {
        private QuizDao mAsyncTaskDao;

        updateQuestionsAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(Long... params) {
            mAsyncTaskDao.updateQuestions(params[0], System.currentTimeMillis());
            return null;
        }
    }

    public void addWord(long quizId, String wordString, String pinyinString) {
        WordItem wordItem = new WordItem(quizId, wordString, pinyinString);
        new addWordAsyncTask(mQuizDao).execute(wordItem);
    }

    private static class addWordAsyncTask extends AsyncTask<WordItem, Void, Void> {
        private QuizDao mAsyncTaskDao;

        addWordAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(WordItem... params) {
            String pinyinString = params[0].getPinyinString();
            String wordString = params[0].getWordString();
            mAsyncTaskDao.insert(new Word(wordString, pinyinString));
            mAsyncTaskDao.insert(new QuizPinyin(params[0].getQuizId(), pinyinString, wordString));
            return null;
        }
    }

    // Only delete QuizPinyin object.
    public void deleteWord(QuizPinyin quizPinyin) {
        new deleteWordAsyncTask(mQuizDao).execute(quizPinyin);
    }

    private static class deleteWordAsyncTask extends AsyncTask<QuizPinyin, Void, Void> {
        private QuizDao mAsyncTaskDao;

        deleteWordAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(QuizPinyin... params) {
            // Word and Pinyin are immutable. Only QuizPinyin object is deleted.
            mAsyncTaskDao.deleteQuizPinyin(params[0].getQuizId(), params[0].getPinyinString());
            return null;
        }
    }

    // Undo a just deleted word.
    public void insertQuizPinyin(QuizPinyin quizPinyin) {
        new insertQuizPinyinAsyncTask(mQuizDao).execute(quizPinyin);
    }

    private static class insertQuizPinyinAsyncTask extends AsyncTask<QuizPinyin, Void, Void> {
        private QuizDao mAsyncTaskDao;

        insertQuizPinyinAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(QuizPinyin... params) {
            mAsyncTaskDao.insert(params[0]);
            return null;
        }
    }

    public void deleteQuiz(long quizId) {
        new deleteQuizItemAsyncTask(mQuizDao).execute(quizId);
    }

    private static class deleteQuizItemAsyncTask extends AsyncTask<Long, Void, Void> {
        private QuizDao mAsyncTaskDao;

        deleteQuizItemAsyncTask(QuizDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(Long... params) {
            mAsyncTaskDao.deleteQuiz(params[0]);
            return null;
        }
    }
}

File: app/src/main/java/com/zhiyong/tingxie/QuizRepository.kt
package com.zhiyong.tingxie

import android.content.Context
import com.google.firebase.auth.FirebaseAuth
import com.zhiyong.tingxie.db.*
import com.zhiyong.tingxie.network.*
import com.zhiyong.tingxie.ui.friend.individual.TingXieIndividual
import com.zhiyong.tingxie.ui.hsk.words.HskWordsAdapter
import com.zhiyong.tingxie.ui.word.WordItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.collections.HashSet


/* A Repository is a class that abstracts access to multiple data sources.
A Repository manages query threads and allows you to use multiple backends.
In the most common example, the Repository implements the logic for deciding whether to fetch data
from a network or use results cached in the local database. */
class QuizRepository(val context: Context) {
  // todo: val database: PinyinRoomDatabase instead of val context: Context.
  private val mQuizDao: QuizDao = getDatabase(context).pinyinDao

  private val executor: ExecutorService = Executors.newSingleThreadExecutor()

  private lateinit var email: String
  private lateinit var name: String

  init {
    // 12/8/22. Should use the user details in the service layer rather than the
    // repository layer, but keeping this here because there is so much code that uses it
    // already.
    val user = FirebaseAuth.getInstance().currentUser
    val email = user?.email
    val name = user?.displayName
    if (email == null || name == null) {
      // todo Log Crashlytics
    } else {
      this.email = email
      this.name = name
    }
  }

  suspend fun putToken(uid: String, email: String, token: String) {
    withContext(Dispatchers.IO) {
      TingXieNetwork.tingxie.putToken(
        NetworkToken(
          uid,
          email,
          token
        )
      )
    }
  }

  suspend fun getQuizzes(): List<NetworkQuiz> = TingXieNetwork.tingxie.getQuizzes(email)

  suspend fun getWordItemsOfQuiz(quizId: Long): List<WordItem> =
    TingXieNetwork.tingxie.getWordItemsOfQuiz(quizId)
      .map { it.asDomainModel(quizId) }

  suspend fun notCorrectWordsRandomOrder(quizId: Long): List<NetworkWordItem> =
    TingXieNetwork.tingxie.notCorrectWordsRandomOrder(quizId, email)

  private var hskWordMap: MutableMap<Int, List<HskWordsAdapter.HskWord>> = mutableMapOf()

  init {
    for (level in 1..6) {
      val wordArray = JSONArray(context.assets
        .open("hsk-vocab-json/hsk-level-$level.json").bufferedReader().use {
          it.readText()
        }
      )
      val hskWords: MutableList<HskWordsAdapter.HskWord> = mutableListOf()
      for (i in 0 until wordArray.length()) {
        val word = wordArray.getJSONObject(i)
        val id = word.getInt("id")
        val hanzi = word.getString("hanzi")
        val pinyin = word.getString("pinyin")
        hskWords.add(HskWordsAdapter.HskWord(id, hanzi, pinyin))
      }
      hskWordMap[level] = hskWords
    }
  }

  fun getHsk(level: Int): List<HskWordsAdapter.HskWord> {
    return hskWordMap[level]!!
  }

  fun getUnaskedHskWords(level: Int): List<HskWordsAdapter.HskWord> {
    val wordList = getHsk(level)
    val unaskedIds = HashSet(wordList.map { word -> word.index })
    // todo: Deduplicate this shared preferences code.
    val sharedPreferences = context.getSharedPreferences("hsk", Context.MODE_PRIVATE)
    val askedIds = HashSet(
      sharedPreferences.getStringSet("askedHskIds", setOf())!!.map { id -> id.toInt() })
    unaskedIds.removeAll(askedIds)
    return wordList.filter { it.index in unaskedIds }
  }

  fun getHanzis(level: Int, pinyin: String): List<String> {
    val wordList = getHsk(level)
    return wordList.filter { it.pinyin == pinyin }.map { it.hanzi }
  }

  fun resetAsked(level: Int): Boolean {
    val wordList = getHsk(level)
    val ids = HashSet(wordList.map { word -> word.index })
    val sharedPreferences = context.getSharedPreferences("hsk", Context.MODE_PRIVATE)
    val askedIds = HashSet(
      sharedPreferences.getStringSet("askedHskIds", setOf())!!
        .filter { it.toInt() !in ids })
    val editor = sharedPreferences.edit()
    editor.putStringSet("askedHskIds", askedIds)
    return editor.commit()
  }

  fun setAsked(index: Int) {
    val sharedPreferences = context.getSharedPreferences("hsk", Context.MODE_PRIVATE)
    val askedIds = HashSet(
      sharedPreferences.getStringSet("askedHskIds", setOf())!!.map { id -> id.toInt() })
    askedIds.add(index)
    val editor = sharedPreferences.edit()
    editor.putStringSet("askedHskIds", HashSet(askedIds.map { id -> id.toString() }))
    editor.apply()
  }

  suspend fun createQuiz(title: String, date: Int): Long {
    return TingXieNetwork.tingxie.postQuiz(NetworkCreateQuiz(title, date, name, email))
  }

  suspend fun updateQuiz(quiz: NetworkQuiz): Int {
    // todo: REPAIR THIS.
    return TingXieNetwork.tingxie.putQuiz(quiz)
//      NetworkQuiz(
//        quiz.id,
//        quiz.title,
//        quiz.date,
//        email,
//        "STUB. REPLACE THIS",
//        quiz.numWords,
//        quiz.numNotCorrect,
//        quiz.round
//      )
//    )
  }

  suspend fun addWord(quizId: Long, wordString: String?, pinyinString: String?): Long {
    if (wordString == null || pinyinString == null) {
      throw IllegalArgumentException("Null wordString or pinyinString.")
    }
    return TingXieNetwork.tingxie.postWord(
      quizId,
      NetworkCreateWord(wordString, pinyinString)
    )
  }

  suspend fun deleteWord(id: Long) = TingXieNetwork.tingxie.deleteWord(id)

  // Undo a just deleted word.
  fun insertQuizPinyin(quizPinyin: QuizPinyin?) = executor.execute {
    mQuizDao.insert(quizPinyin)
  }

  suspend fun deleteQuiz(quizId: Long) =
    TingXieNetwork.tingxie.deleteQuiz(quizId, name, email, email)

  suspend fun upsertCorrectRecord(wordId: Long) =
    TingXieNetwork.tingxie.upsertCorrectRecord(NetworkCorrectRecord(wordId, email))

  suspend fun getGroups(): MutableList<NetworkGroup> =
    TingXieNetwork.tingxie.getGroups(email)

  suspend fun createGroup(groupName: String, members: List<NetworkGroupMember>): String =
    TingXieNetwork.tingxie.createGroup(
      NetworkCreateGroup(groupName, name, email, members)
    )

  suspend fun deleteGroup(
    groupId: Long,
    requesterName: String,
    requesterEmail: String
  ): Int = TingXieNetwork.tingxie.deleteGroup(groupId, requesterName, requesterEmail)

  suspend fun getGroupMembers(groupId: Long) =
    TingXieNetwork.tingxie.getGroupMembers(groupId)

  suspend fun addGroupMemberOrReturnNoUser(
    groupId: Long,
    email: String,
    role: String
  ): NetworkGroupMember =
    TingXieNetwork.tingxie.addGroupMemberOrReturnNoUser(groupId, email, role)

  suspend fun changeRole(groupId: Long, email: String, role: String): Int =
    TingXieNetwork.tingxie.changeRole(groupId, email, role)

  suspend fun deleteGroupMember(
    groupId: Long,
    requesterName: String,
    requesterEmail: String,
    email: String
  ): Int = TingXieNetwork.tingxie.deleteGroupMember(
    groupId,
    email,
    requesterName,
    requesterEmail
  )

  suspend fun getUsersOfQuiz(quizId: Long): List<NetworkGroupMember> =
    TingXieNetwork.tingxie.getUsersOfQuiz(quizId)

  suspend fun addQuizMemberOrReturnNoUser(
    quizId: Long,
    addQuizUser: NetworkAddQuizUser
  ): NetworkGroupMember =
    TingXieNetwork.tingxie.addQuizMemberOrReturnNoUser(quizId, addQuizUser)

  suspend fun removeQuizMember(
    quizId: Long,
    requesterName: String,
    requesterEmail: String,
    email: String
  ): String = TingXieNetwork.tingxie.deleteQuiz(quizId, requesterName, requesterEmail, email)


  // Endpoints below are for Friends API, which I decided not to use.
  suspend fun checkUserExists(email: String): Boolean =
    TingXieNetwork.tingxie.checkUserExists(email).toBoolean()

  suspend fun getFriends(party: String, friendStatus: String): List<TingXieIndividual> =
    TingXieNetwork.tingxie.getFriends(email, party, friendStatus)
      .map { it.asDomainModel() }

  suspend fun addFriend(individual: TingXieIndividual) =
    TingXieNetwork.tingxie.postFriend(
      NetworkIndividual(
        email,
        individual.email,
        name,
        individual.name,
        individual.status
      )
    )

  suspend fun updateFriend(individual: TingXieIndividual) =
    TingXieNetwork.tingxie.putFriend(
      NetworkIndividual(
        individual.email,
        email,
        name,
        individual.name,
        individual.status
      )
    )

  suspend fun deleteFriend(email: String) =
    TingXieNetwork.tingxie.deleteFriend(this.email, email)

//  suspend fun acceptOtherIndividualRequest(email: String) =
//    TingXieNetwork.tingxie.putOtherIndividualRequest(email, this.email, true)
//
//  suspend fun rejectOtherIndividualRequest(email: String) =
//    TingXieNetwork.tingxie.putOtherIndividualRequest(email, this.email, false)

//  suspend fun getShares(quizId: Long): List<TingXieShareIndividual> =
//    TingXieNetwork.tingxie.getShares(email, quizId).map {
//      TingXieShareIndividual(it.email, it.name, it.isShared, it.role)
//    }
//    return arrayListOf(
//        TingXieShareIndividual("yongish@gmail.com", "firstZ", "lastZ", true, EnumQuizRole.EDITOR),
//        TingXieShareIndividual("test0email.com", "first0", "last0", false, EnumQuizRole.EDITOR),
//        TingXieShareIndividual("test1email.com", "first1", "last1", false, EnumQuizRole.VIEWER),
//        TingXieShareIndividual("test2email.com", "first2", "last2", true, EnumQuizRole.VIEWER),
//    )
//    return arrayListOf()

//  suspend fun addShare(quizId: Long, shareIndividual: TingXieShareIndividual) {
//    TingXieNetwork.tingxie.postShare(
//      email, quizId, NetworkShare(
//        shareIndividual.email, shareIndividual.name, true, shareIndividual.role
//      )
//    )
//  }

//  suspend fun shareAll(quizId: Long, shared: Boolean): List<TingXieShareIndividual> {
//    TingXieNetwork.tingxie.putShareAll(email, quizId, shared)
//    return arrayListOf()
//  }

  suspend fun deleteShare(quizId: Long, email: String) {
    TingXieNetwork.tingxie.deleteShare(this.email, quizId, email)
  }

  companion object {
    private const val TAG = "QuizRepository"
  }
}